cupy/cuda/memory_hook.py

```diff
@@ -0,0 +1,96 @@
+import collections
+import threading
+
+thread_local = threading.local()
+
+
+def get_memory_hooks():
+    if not hasattr(thread_local, 'memory_hooks'):
+        thread_local.memory_hooks = collections.OrderedDict()
+    return thread_local.memory_hooks
+
+
+class MemoryHook(object):
+    """Base class of hooks for Memory allocations.
+
+    :class:`~cupy.cuda.MemoryHook` is an callback object
+    that is registered to :class:`~cupy.cuda.SingleDeviceMemoryPool`.
```

Comment:
It seems that instances of `SingleDeviceMemoryPool` of different device IDs use same set of `MemoryHook`s as long as they are handled in the same thread. Is it intentional?